databuilder/extractor/bigquery_usage_extractor.py

```diff
@@ -0,0 +1,103 @@
+from collections import namedtuple
+from datetime import date, timedelta
+import logging
+
+from googleapiclient.discovery import build
+from oauth2client.client import GoogleCredentials
+
+from databuilder.extractor.base_extractor import Extractor
+
+TableColumnUsageTuple = namedtuple('TableColumnUsageTuple', ['database', 'cluster', 'schema',
+                                                             'table', 'column', 'email'])
+
+LOGGER = logging.getLogger(__name__)
+
+
+class BigQueryUsageExtractor(Extractor):
+    """
+    An aggregate extractor for bigquery table usage. This class takes the data from
+    the stackdriver logging API by filtering on timestamp, bigquery_resource and looking
+    for referencedTables in the response.
+    """
+    TIMESTAMP_KEY = 'timestamp'
+    PROJECT_ID_KEY = 'projectid'
+
+    def init(self, conf):
+        credentials = GoogleCredentials.get_application_default()
+        self.logging_service = build('logging', 'v2', credentials=credentials)
+
+        # type: (ConfigTree) -> None
+        self.timestamp = conf.get_string(
+            BigQueryUsageExtractor.TIMESTAMP_KEY,
+            (date.today() - timedelta(days=1)).strftime('%Y-%m-%dT00:00:00Z'))
+        self.projectid = conf.get_string(BigQueryUsageExtractor.PROJECT_ID_KEY)
+        self.table_usage_counts = {}
+        self._count_usage()
+        self.iter = iter(self.table_usage_counts)
+
+    def _count_usage(self):
+        for entry in self._retrieve_records():
+            if not isinstance(entry, dict):
+                raise Exception("BigQueryUsageExtractor expects record of type dict!")
+
+            email = entry['protoPayload']['authenticationInfo']['principalEmail']
+            jobCompletedEvent = entry['protoPayload']['serviceData']['jobCompletedEvent']
+            refTables = jobCompletedEvent['job']['jobStatistics'].get('referencedTables', None)
+
+            if not refTables:
```

Comment:
I don't think I understand what this means for usage count. When `refTables` will be `None`?

Reply:
From my understanding on the code, it seems everytime the table has been referenced once that consider one usage.

Reply:
See above comment, BQ will only actually run a query when it can't use a cached result run previously (available for up to 24 hours).

Reply:
Still trying to connect the dot between caching behavior and usage count, and please bear with me.
Does this mean that `.get('referencedTables')` will provide the table if it's not from cached result? For example, if same query is being issued multiple times within 24 hours resulting cache hit, it will still count as one? Furthermore, would BQ extend TTL on the cache so that it can extend more than 24 hours?

Reply:
@gtoonstra in addition to @jinhyukchang's question, would you happen to know what happens if the same query is issued my two different users? Would BQ use the cached result in this case (thus resulting in one usage count) ?

Reply:
@gtoonstra Gentle ping when you have a chance.

Reply:
sorry, will answer tonight or tomorrow.

Reply:
@jinhyukchang : Yes, only when the query result was not cached at the time will the referencedTables be in there. Under the hood what happens in BQ is that when you run a query, it will create a temporary BigQuery table for the user. This table exists for up to 24 hours. So what I interpret from the observations is that as long as the cache was used instead of touching the original tables, it will not have referencedTables in there because they weren't actually accessed. You can turn the cache on/off in the browser, but not change the TTL. Yes, if the cache is hit, refTables are not populated, so it will count as one. @Alagappan : Cache results are per user per project: https://cloud.google.com/bigquery/docs/cached-results , so cache results are not shared between users.